hbase-server/src/main/java/org/apache/hadoop/hbase/master/ServerManager.java

```diff
@@ -163,6 +169,44 @@
   private final ConcurrentNavigableMap<ServerName, ServerMetrics> onlineServers =
     new ConcurrentSkipListMap<>();
 
+  /**
+   * Store the snapshot of the current region server list, for improving read performance.
+   * <p/>
+   * The hashCode is used to determine whether there are changes to the region servers.
+   */
+  private static final class OnlineServerListSnapshot {
+
+    private static final HashFunction HASH = Hashing.murmur3_128();
+
+    final List<ServerName> servers;
+
+    final long hashCode;
+
+    public OnlineServerListSnapshot(List<ServerName> servers) {
+      this.servers = Collections.unmodifiableList(servers);
+      Hasher hasher = HASH.newHasher();
+      for (ServerName server : servers) {
+        hasher.putString(server.getServerName(), StandardCharsets.UTF_8);
+      }
+      this.hashCode = hasher.hash().asLong();
```

Comment:
Does hash conflict need to be considered? Although the probability is small.
And if a server go online and offline between a client two calls, the hashCode is same for the client, is this expected?

Reply:
Ah, this is a problem. In real world, for the replication scenario, usually it is not a problem as it is not likely that a regionserver which can only live for less than 5 seconds could hold any regions. But consider general usage, maybe we should let users know there is a dead server.

Let me think how to deal with this.

Reply:
And this could also be a problem for the old zk based implementation. We just watch znode and process the znode deleted event, but it is also possible that, after a trigger of the watcher, before set a new watcher, a region server is up and then down, then you can not get the deleted event either.

So the solution should be general, maybe we need to get all the replicators from the replication queue storage, and compare it with the current live region server set, to find out the dead region server.